sparkles/roll_optimize.py

```diff
@@ -178,16 +178,19 @@ def get_ids_list(roll_offsets):
         if roll_dev is None:
             roll_dev = allowed_rolldev(pitch)
 
-        # Ensure roll_nom in range 0 <= roll_nom < 360 to match q_att.roll
+        # Ensure roll_nom in range 0 <= roll_nom < 360 to match q_att.roll.
+        # Also ensure that roll_min < roll < roll_max.  It can happen that the
+        # ORviewer scheduled roll is outside the allowed_rolldev() range.  For
+        # far-forward sun, allowed_rolldev() = 0.0.
+        roll = q_att.roll
         roll_nom = roll_nom % 360.0
-        roll_min = roll_nom - roll_dev
-        roll_max = roll_nom + roll_dev
+        roll_min = min(roll_nom - roll_dev, roll - 0.1)
+        roll_max = max(roll_nom + roll_dev, roll + 0.1)
```

Comment:
And these look good to me but I'm not sure about the 0.1 values.

Reply:
My concern with the 0.1 values was just that for d_roll = 0.25, I didn't know if you could get into a situation with, say at the positive case, roll - roll_nom = 0.16, and then end up with the range for a zero roll_dev containing roll_nom, roll, *and* an extra candidate roll.  Given that the point of the tool is to provide rolls to check that are already advertised as experimental and perhaps not allowed, this seems fine (especially for the weird case where roll_dev is 0 and roll is a good chunk (> 0.15) from roll_nom etc).

Reply:
The case that actually caused the problem had roll - roll_nom of around 0.34.  As you say, there is really no concern with extra candidate rolls, in fact the fundamental design is always to sample rolls that are outside of the allowed range.  In fact another thought I had for fixing things was basically:
```
        padding = roll_dev + d_roll * 2  # Make sure there are at least a few samples
        ro_minus = np.arange(0, roll_min - padding - roll, -d_roll)[1:][::-1]
        ro_plus = np.arange(0, roll_max + padding - roll, d_roll)
```
But at the end this won't really help because the allowed options will be clipped by roll_min and roll_max.  And it *is important* to make sure that `roll_min <= roll <= roll_max`, so the simpler fix of just opening up roll_min and roll_max seemed better.

Reply:
Regarding "And it _is important_ to make sure that `roll_min <= roll <= roll_max`, so the simpler fix of just opening up roll_min and roll_max seemed better.",  right, I was trying to wrap my head around any other implicit requirements.

Reply:
I know this is still a beta feature, but I am unclear about which piece of data is "real" or "more important".  If the `roll_min`/`roll_max` can be (arbitrarily) expanded, then is the reported `roll` the "real" thing that needs to be safeguarded?

Another way to put this:

It seems, from the description of the issue, that this could be fixed by either expanding the `roll_min` - `roll_max` range to always include the `roll`, **or** by moving the `roll` to always be inside the `roll_min` - `roll_max` range.

Why is expanding the range the "right" choice?

Reply:
If I'm understanding correctly, in this case the as-scheduled roll (aka "roll") was outside the tool's calculated range.  So I was with Tom in that if the mission planners supply knowledge somewhat counter to the tool, it is relatively reasonable to just go along with that instead of returning a "your observation was outside the roll range we've calculated (apparently badly)".

Reply:
ah, that makes sense.  I was making the assumption that the roll is usually being chosen by proseco or based on some estimation of what the roll should be, not coming from the mission planner.  I agree that if the roll is provided by the mission planner, it would likely be best to not modify it.

Reply:
And I still said it badly, but I think we're on the same page 👍

Reply:
Agreed with what @jeanconn said.